lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+	case crypto.SHA256, 0:
+		alg = kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_2048_SHA256
+	case crypto.SHA512:
+		alg = kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_4096_SHA512
+	default:
+		return nil, nil, trace.BadParameter("unsupported digest algorithm: %v", options.DigestAlgorithm)
+	}
+
+	keyUUID, err := newKeyID()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	req := &kmspb.CreateCryptoKeyRequest{
+		Parent:      g.keyRing,
+		CryptoKeyId: keyUUID,
+		CryptoKey: &kmspb.CryptoKey{
+			Purpose: kmspb.CryptoKey_ASYMMETRIC_SIGN,
+			Labels: map[string]string{
+				hostLabel: g.hostUUID,
+			},
+			VersionTemplate: &kmspb.CryptoKeyVersionTemplate{
+				ProtectionLevel: g.protectionLevel,
+				Algorithm:       alg,
+			},
+		},
+	}
+	resp, err := doGCPRequest(ctx, g.clock, g.kmsClient.CreateCryptoKey, req)
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "error while attempting to generate new GCP KMS key")
+	}
+
+	keyID := gcpKMSKeyID{
+		keyVersionName: resp.Name + keyVersionSuffix,
+	}
+
+	signer, err := g.newKmsSigner(ctx, keyID)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return keyID.marshal(), signer, nil
+}
+
+// getSigner returns a crypto.Signer for the given pem-encoded private key.
+func (g *gcpKMSKeyStore) getSigner(rawKey []byte) (crypto.Signer, error) {
+	ctx := context.TODO()
+	keyID, err := parseGCPKMSKeyID(rawKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	signer, err := g.newKmsSigner(ctx, keyID)
+	return signer, trace.Wrap(err)
+}
+
+// deleteKey deletes the given key from the KeyStore.
+func (g *gcpKMSKeyStore) deleteKey(ctx context.Context, rawKey []byte) error {
+	keyID, err := parseGCPKMSKeyID(rawKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	req := &kmspb.DestroyCryptoKeyVersionRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry to destroy while the key is still pending creation.
+	_, err = retryWhilePending(ctx, g.clock, g.kmsClient.DestroyCryptoKeyVersion, req)
+	return trace.Wrap(err, "error while attempting to delete GCP KMS key")
+}
+
+// canSignWithKey returns true if given a GCP_KMS key in the same key ring
+// managed by this keystore. This means that it's possible (and expected) for
+// multiple auth servers in a cluster to sign with the same KMS keys if they are
+// configured with the same keyring. This is a divergence from the PKCS#11
+// keystore where different auth servers will always create their own keys even
+// if configured to use the same HSM
+func (g *gcpKMSKeyStore) canSignWithKey(raw []byte, keyType types.PrivateKeyType) (bool, error) {
+	if keyType != types.PrivateKeyType_GCP_KMS {
+		return false, nil
+	}
+	keyID, err := parseGCPKMSKeyID(raw)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if !strings.HasPrefix(keyID.keyVersionName, g.keyRing) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// DeleteUnusedKeys deletes all keys from KMS if they are:
+// 1. Labeled by this server (matching HostUUID) when they were created
```

Comment:
Do keys have an expiration date? How do we deal with setups that (rightfully?) have ephemeral auth servers? Just don't support those?

Reply:
Ephemeral auth servers will actually be supported, they can use keys created by other auth servers perfectly fine, it's just that they won't delete any old keys created by other auth servers. The admin will be able to delete those manually. The reason is to avoid deleting any keys created by a different auth server that haven't been added to the CA yet. Also if they are running multiple clusters with the same keyring, or using other keys in that keyring. Could label with the cluster name, but that does not avoid the first problem of 2 auth servers in 1 cluster deleting each others keys.

Reply:
The keys do not have an expiration date